Infer type from field instead of column

armenio opened the pull request:
getTypeOfColumn() relies on getTypeOfField(), and does not suffer from
mismatching issues caused by quoting, because you cannot quote a field.
Since a field can be composite, that method returns an array, hence why we
need to select the first element.

Fixes https://github.com/doctrine/orm/issues/9109 caused by https://github.com/doctrine/orm/pull/9010

armenio commented:
need labels: "Bug" and "Failing Test"

greg0ire commented:
>  In attempt to resolve doctrine#9109 

Please improve your commit message according to [the contributing guide](https://www.doctrine-project.org/contribute/index.html#working-on-topic-branches).

Specifically, the commit message should explain what did not work before exactly, and why it works now.

greg0ire commented on tests/Doctrine/Tests/ORM/Functional/Ticket/GH9109Test.php:
```diff
@@ -0,0 +1,201 @@
+                $this->_em->getClassMetadata(GH9109User::class),
+                $this->_em->getClassMetadata(GH9109Product::class),
+            ]
+        );
```
Please call `$this->_schemaTool->dropSchema(…)` in `tearDown()`

armenio replied:
done

greg0ire commented on tests/Doctrine/Tests/ORM/Functional/Ticket/GH9109Test.php:
```diff
@@ -0,0 +1,201 @@
+    }
+
+    /**
+     * @return Collection
```
Please use annotations similar to those line 112 - 113

armenio replied:
done

greg0ire commented on tests/Doctrine/Tests/ORM/Functional/Ticket/GH9109Test.php:
```diff
@@ -0,0 +1,201 @@
+
+        // assert Product has a Buyer
+        $criteria = Criteria::create();
+        $count = $persistedProduct->getBuyers()->matching($criteria)->count();
```
What's the point of calling `matching` with an empty criteria?

armenio replied:
fixed

greg0ire commented on tests/Doctrine/Tests/ORM/Functional/Ticket/GH9109Test.php:
```diff
@@ -0,0 +1,201 @@
+        // assert NOT QUOTED WORK with Criteria
+        $criteria = Criteria::create();
+        $criteria->where($criteria->expr()->eq('lastName', $userLastName));
+        $matchedUserLastName = $persistedProduct->getBuyers()->matching($criteria)->first();
```
This variable name is confusing: one might think it will be a name, when it really is a user. IMO you could just reuse `$user` here, like you did below

armenio replied:
fixed. Thanks for the suggestion

greg0ire commented on tests/Doctrine/Tests/ORM/Functional/Ticket/GH9109Test.php:
```diff
@@ -0,0 +1,201 @@
+        // assert QUOTED DOES WORK with Criteria
+        $criteria = Criteria::create();
+        $criteria->where($criteria->expr()->eq('firstName', $userFirstName));
+        $matchedUserFirstName = $persistedProduct->getBuyers()->matching($criteria)->first();
```
Same here.

armenio replied:
fixed. Thanks for the suggestion

greg0ire commented:
Please kindly squash your commits together. If you don't, we'll try to remember to do it for you but it's best if you save us this trouble.

How to do that?

1. `git rebase -i origin/2.10.x`, assuming `origin` is a git remote that points to this repository, and not your fork. If you're not sure what your remotes are, run `git remote -vvv`, there should be your fork and the holy/reference/base/origin/whatever-you-call-it repository.
2. A window will show up with many lines, replace `pick` with `fixup` on every line but the first one
3. For the first line, use `r`, so that you get a chance to re-use the commit message in https://github.com/doctrine/orm/pull/9153#issuecomment-961661495
4. Close your editor, git should do its magic, and you should end up with one commit
5. Use `git push --force` to overwrite what you already push. Don't forget the `--force` option otherwise git will try to merge both things together.

armenio commented:
Commits are squashed

greg0ire commented:
Please use `git commit --amend` to drop the `In attempt to resolve #9109` part of your commit message.

armenio committed:
Infer type from field instead of column

getTypeOfColumn() relies on getTypeOfField(), and does not suffer from
mismatching issues caused by quoting, because you cannot quote a field.
Since a field can be composite, that method returns an array, hence why we
need to select the first element.

armenio commented:
done.

greg0ire reviewed (dismissed):
As a related but separate issue from the issue at hand, I think we should either document that the `$columnName` argument is supposed to be quoted if that's indeed the intended usage, or improve that method to make it work regardless of whether it is quoted or not.

armenio commented:
Thank you very much for your effort @greg0ire

greg0ire commented:
@armenio you're welcome! I notice there is an issue with PostgreSQL though, can you please look into it?

armenio commented:
Fixed PgSQL error with changing strategy to "AUTO" in @GeneratedValue of PK's

beberlei commented:
@greg0ire This should maybe even target a new 2.9.7 or? It was merged to 2.9.6 as far as i see breaking a core function.

greg0ire commented:
@beberlei 2.9.x is no longer supported, and upgrading to 2.10.x does not seem like a huge deal so I would backport it iff someone complains they cannot upgrade. Otherwise, it's a good incentive to get people to upgrade IMO 🙂

greg0ire commented:
Thanks @armenio !
